### PR TITLE
Add a TracingBuilder for library instrumentation to extend from for c…

### PR DESCRIPTION
**Pull request by anuraaga**
…ommon configuration.

We'd like library instrumentation to expose a consistent set of configuration to users, and this makes sure that is automatic.

Sending for an initial look to make sure it generally looks like a sane approach.

### CODE REVIEW

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/InstrumenterBuilder.java [hunk 1]
```diff
@@ -199,6 +199,17 @@
         InstrumenterConstructor.propagatingToDownstream(setter), SpanKindExtractor.alwaysClient());
   }
 
+  /**
+   * Returns a new {@link Instrumenter} which will create client spans and inject context into
+   * requests.
+   */
+  public Instrumenter<REQUEST, RESPONSE> newClientInstrumenter(
+      TracingBuilder<REQUEST, RESPONSE, ?> tracingBuilder, TextMapSetter<REQUEST> setter) {
```
Comment (anuraaga) [inline instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/InstrumenterBuilder.java:207]: The idea is to remove the methods that don't accept a `TracingBuilder`

Comment (mateuszrzeszutek) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/InstrumenterBuilder.java:207]: Should we really do that? Requiring a `TracingBuilder` will make it harder to use `Instrumenter` ad-hoc - for example in our own javaagent instrumentations (or in libs like spring sleuth, probably).

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/TracingBuilder.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+package io.opentelemetry.instrumentation.api.instrumenter;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Function;
+
+public abstract class TracingBuilder<
```
Comment (mateuszrzeszutek) [inline instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/TracingBuilder.java:14]: This is probably a good opportunity to revisit https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/3156 - should it really be `TracingBuilder`? Maybe `TelemetryBuilder`? (and `FooTelemetryBuilder`, `FooTelemetry`)

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/InstrumenterBuilder.java (hunk 1 above)
Comment (anuraaga) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/InstrumenterBuilder.java:207]: Good point. I struggled to come up with a pattern for applying the settings to the Instrumenter that isn't easy to forget to do, e.g. a protected `apply` method.

Any idea come to mind?

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/TracingBuilder.java (hunk 2 above)
Comment (anuraaga) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/TracingBuilder.java:14]: Yeah kept this consistent with what we have for now but realized we should rename to something like FooInstrumentation

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/InstrumenterBuilder.java (hunk 1 above)
Comment (mateuszrzeszutek) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/InstrumenterBuilder.java:207]: I mean, adding new `newInstrumenter()` methods that accept a `TracingBuilder` seems fine, it's the removing idea that might be problematic; I don't really have better ideas that this. `copySettingsFrom(TracingBuilder)`?

**Comment by trask** [comment]
@mateuszrzeszutek @anuraaga is there anything from this PR that we should consider as part of stabilizing the instrumenter api?

**Comment by breedx-splk** [comment]
Closing a few old/stale draft PRs. Reopen if interested.

**Pull request closed without merging**